state-chain/chains/src/sol/sol_tx_core/address_derivation.rs

```diff
@@ -0,0 +1,149 @@
+//! Contains code for Deriving addresses and deposit channels for the Solana Environment.
+use cf_primitives::{chains::assets::sol::Asset as SolAsset, ChannelId};
+use sol_prim::AccountBump;
+
+use crate::{
+	deposit_channel::DepositChannel,
+	sol::{consts::*, AddressDerivationError, DerivedAddressBuilder, SolAddress},
+	Solana,
+};
+use core::str::FromStr;
+
+/// Derive address/deposit channels for a channel ID
+pub fn derive_deposit_channel(
+	channel_id: ChannelId,
+	_asset: SolAsset,
+	vault_program: SolAddress,
+) -> Result<DepositChannel<Solana>, AddressDerivationError> {
+	let seed = channel_id.to_le_bytes();
+	derive_address(seed, vault_program)
+		// .and_then(|(address, bump)| match asset {
+		// 	SolAsset::Sol => Ok((address, bump)),
+		// 	SolAsset::Usd => {
+		// 	    let mint_pubkey =
+		// 	Env::lookup_account(crate::sol::api::SolanaEnvAccountLookupKey::TokenMintPubkey)?;
+		// 	    derive_associated_token_account(address, mint_pubkey)
+		// 	},
+		// })
+		.map(|(address, bump)| DepositChannel::<Solana> {
+			channel_id,
+			address,
+			asset: SolAsset::Sol,
+			state: bump,
+		})
+}
+
+/// Derive an address from our Vault program key. Produces an Address and a bump.
+fn derive_address(
+	seed: impl AsRef<[u8]>,
+	vault_program: SolAddress,
+) -> Result<(SolAddress, AccountBump), AddressDerivationError> {
+	DerivedAddressBuilder::from_address(vault_program)?.chain_seed(seed)?.finish()
+}
+
+#[allow(dead_code)]
+/// Derive a Associated Token Account (ATA) of a main target account.
+fn derive_associated_token_account(
+	target: SolAddress,
+	mint_pubkey: SolAddress,
+) -> Result<(SolAddress, AccountBump), AddressDerivationError> {
+	let associated_token_program_id = SolAddress::from_str(ASSOCIATED_TOKEN_PROGRAM_ID)
```

Comment:
Not related specifically to this PR, but we have a lot of constants that are stored as a string and we convert them at runtime - there is probably a way to do the conversion at compile time...

Reply:
Update: it looks like bs58 allows this - we just need to bump the version from 0.4 to 0.5.1. (we can consider this for later). This would have the advantage of compile-time checking all the base58 strings to ensure they are valid base58 of the correct length.

I just tested this locally:

```rust
const fn const_pubkey(s: &'static str) -> Pubkey {
	Pubkey(bs58::decode(s.as_bytes()).into_array_const_unwrap())
}

// This is fine
const VALID_PUBKEY: Pubkey = const_pubkey("274BzCz5RPHJZsxdcSGySahz4qAWqwSDcmz1YEKkGaZC");

// This causes a compile-time error
const INVALID_PUBKEY: Pubkey = const_pubkey("--invalid chars--");
```

Reply:
I added an issue to Linear.